# 003-cache-sync: Cache thread sync/no-thread: 6 caches converted to mutex-protected or no-thread mode
# Files: 12
# Generated by scripts/split-consolidated-patches.py
diff --git a/Source/Reasoner/Kernel/Cache/CComputedConsequencesCache.cpp b/Source/Reasoner/Kernel/Cache/CComputedConsequencesCache.cpp
index 8ee0a748..42eb1dd0 100644
--- a/Source/Reasoner/Kernel/Cache/CComputedConsequencesCache.cpp
+++ b/Source/Reasoner/Kernel/Cache/CComputedConsequencesCache.cpp
@@ -43,7 +43,9 @@ namespace Konclude {
 						mConfAllowedNonDetExpansionCount = CConfigDataReader::readConfigInteger(config,"Konclude.Calculation.Optimization.SaturationExpansionSatisfiabilityCacheCount",1);
 					}
 
+#ifndef WASM_BUILD
 					startThread(QThread::HighestPriority);
+#endif
 				}
 
 
@@ -53,7 +55,14 @@ namespace Konclude {
 
 
 				CComputedConsequencesCache* CComputedConsequencesCache::writeCacheData(CComputedConsequencesCacheWriteData* writeData, CMemoryPool* memoryPools) {
+#ifdef WASM_BUILD
+					mWriteSyncMutex.lock();
+					CWriteComputedConcequencesCacheEntryEvent* ev = new CWriteComputedConcequencesCacheEntryEvent(writeData,memoryPools);
+					processCustomsEvents(CWriteComputedConcequencesCacheEntryEvent::EVENTTYPE, ev);
+					mWriteSyncMutex.unlock();
+#else
 					postEvent(new CWriteComputedConcequencesCacheEntryEvent(writeData,memoryPools));
+#endif
 					return this;
 				}
 
diff --git a/Source/Reasoner/Kernel/Cache/CComputedConsequencesCache.h b/Source/Reasoner/Kernel/Cache/CComputedConsequencesCache.h
index b1f25fe8..408e1350 100644
--- a/Source/Reasoner/Kernel/Cache/CComputedConsequencesCache.h
+++ b/Source/Reasoner/Kernel/Cache/CComputedConsequencesCache.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_KERNEL_CACHE_CCOMPUTEDCONSEQUENCESCACHE_H
 
 // Libraries includes
-#include <QHash>
+#include "QtCompat.h"
 #include <QMutex>
 #include <QSemaphore>
 
@@ -121,6 +121,9 @@ namespace Konclude {
 
 						CComputedConsequencesCacheReader* mReaderLinker;
 						QMutex mReaderSyncMutex;
+#ifdef WASM_BUILD
+						QMutex mWriteSyncMutex;
+#endif
 
 						CCacheStatistics mCacheStat;
 						CComputedConsequencesCacheContext mContext;
diff --git a/Source/Reasoner/Kernel/Cache/COccurrenceStatisticsCache.cpp b/Source/Reasoner/Kernel/Cache/COccurrenceStatisticsCache.cpp
index 294c82a7..ffbf0cee 100644
--- a/Source/Reasoner/Kernel/Cache/COccurrenceStatisticsCache.cpp
+++ b/Source/Reasoner/Kernel/Cache/COccurrenceStatisticsCache.cpp
@@ -39,7 +39,9 @@ namespace Konclude {
 					mCacheData = new COccurrenceStatisticsCacheData();
 
 					
+#ifndef WASM_BUILD
 					startThread(QThread::HighestPriority);
+#endif
 				}
 
 
@@ -70,7 +72,14 @@ namespace Konclude {
 
 
 				COccurrenceStatisticsCache* COccurrenceStatisticsCache::writeCachedData(COccurrenceStatisticsCacheWriteData* writeData, CMemoryPool* memoryPools) {
+#ifdef WASM_BUILD
+					mWriteSyncMutex.lock();
+					CWriteCachedDataEvent* ev = new CWriteCachedDataEvent(writeData, memoryPools);
+					processCustomsEvents(CWriteCachedDataEvent::EVENTTYPE, ev);
+					mWriteSyncMutex.unlock();
+#else
 					postEvent(new CWriteCachedDataEvent(writeData, memoryPools));
+#endif
 					return this;
 				}
 
diff --git a/Source/Reasoner/Kernel/Cache/COccurrenceStatisticsCache.h b/Source/Reasoner/Kernel/Cache/COccurrenceStatisticsCache.h
index b3cc955c..4d0e3608 100644
--- a/Source/Reasoner/Kernel/Cache/COccurrenceStatisticsCache.h
+++ b/Source/Reasoner/Kernel/Cache/COccurrenceStatisticsCache.h
@@ -22,8 +22,7 @@
 #define KONCLUDE_REASONER_KERNEL_CACHE_COCCURRENCESTATISTICSCACHE_H
 
 // Libraries includes
-#include <QHash>
-#include <QSet>
+#include "QtCompat.h"
 #include <QMutex>
 #include <QSemaphore>
 
@@ -115,6 +114,9 @@ namespace Konclude {
 
 
 						cint64 mWriteDataCount;
+#ifdef WASM_BUILD
+						QMutex mWriteSyncMutex;
+#endif
 
 
 						CCacheStatistics mCacheStat;
diff --git a/Source/Reasoner/Kernel/Cache/COccurrenceUnsatisfiableCache.cpp b/Source/Reasoner/Kernel/Cache/COccurrenceUnsatisfiableCache.cpp
index 20fb3ec4..74e526cc 100644
--- a/Source/Reasoner/Kernel/Cache/COccurrenceUnsatisfiableCache.cpp
+++ b/Source/Reasoner/Kernel/Cache/COccurrenceUnsatisfiableCache.cpp
@@ -37,8 +37,16 @@ namespace Konclude {
 					writeOperationsCount = 0;
 					mCachingTag = 0;
 					lastUpdateSlot = nullptr;
-
-					startThread();
+					mConfDirectUpdateSynchronization = false;
+#ifdef WASM_BUILD
+					mConfDirectUpdateSynchronization = true;
+#endif
+
+					#ifdef WASM_BUILD
+						threadStarted();
+					#else
+						startThread();
+					#endif
 				}
 
 
@@ -82,16 +90,27 @@ namespace Konclude {
 
 
 				COccurrenceUnsatisfiableCacheEntry *COccurrenceUnsatisfiableCache::getPrimarCacheEntry() {
+#ifndef WASM_BUILD
 					if (!CThread::isThreadRunning()) {
 						CThread::waitSynchronization();
 					}
+#endif
 					return primarCacheEntry;
 				}
 
 
 				void COccurrenceUnsatisfiableCache::addUnsatisfiableCacheEntry(QList<CCacheValue> &itemList) {
 					// item list is and has to be copied in event constructor!
-					CThread::postEvent(new CWriteUnsatisfiableCacheEntryEvent(itemList));
+					if (mConfDirectUpdateSynchronization) {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::OccUnsatCache",logTr("[OccUnsatCache] sync write"),this);
+						mDirectUpdateSyncMutex.lock();
+						CWriteUnsatisfiableCacheEntryEvent* ev = new CWriteUnsatisfiableCacheEntryEvent(itemList);
+						processCustomsEvents(EVENTWRITEUNSATISFIABLECACHEENTRY, ev);
+						mDirectUpdateSyncMutex.unlock();
+					} else {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::OccUnsatCache",logTr("[OccUnsatCache] async write"),this);
+						CThread::postEvent(new CWriteUnsatisfiableCacheEntryEvent(itemList));
+					}
 				}
 
 
diff --git a/Source/Reasoner/Kernel/Cache/COccurrenceUnsatisfiableCache.h b/Source/Reasoner/Kernel/Cache/COccurrenceUnsatisfiableCache.h
index fe7e2503..8a83bf26 100644
--- a/Source/Reasoner/Kernel/Cache/COccurrenceUnsatisfiableCache.h
+++ b/Source/Reasoner/Kernel/Cache/COccurrenceUnsatisfiableCache.h
@@ -22,12 +22,9 @@
 #define KONCLUDE_REASONER_KERNEL_CACHE_COCCURRENCEUNSATISFIABLECACHE_H
 
 // Libraries includes
-#include <QHash>
+#include "QtCompat.h"
 #include <QMutex>
 #include <QSemaphore>
-#include <QVector>
-#include <QList>
-#include <QStringList>
 #include <QReadWriteLock>
 
 // Namespace includes
@@ -168,6 +165,9 @@ namespace Konclude {
 						QString mCachingString;
 						QStringList mCachedStringList;
 
+						bool mConfDirectUpdateSynchronization;
+						QMutex mDirectUpdateSyncMutex;
+
 				};
 
 			}; // end namespace Cache
diff --git a/Source/Reasoner/Kernel/Cache/CReuseCompletionGraphCache.cpp b/Source/Reasoner/Kernel/Cache/CReuseCompletionGraphCache.cpp
index 803c44f0..c4283475 100644
--- a/Source/Reasoner/Kernel/Cache/CReuseCompletionGraphCache.cpp
+++ b/Source/Reasoner/Kernel/Cache/CReuseCompletionGraphCache.cpp
@@ -32,7 +32,9 @@ namespace Konclude {
 
 
 				CReuseCompletionGraphCache::CReuseCompletionGraphCache(QString threadIdentifierName, CWatchDog *watchDogThread) : CThread(threadIdentifierName,watchDogThread) {
+#ifndef WASM_BUILD
 					startThread(QThread::HighestPriority);
+#endif
 
 					mSlotLinker = nullptr;
 					mReaderLinker = nullptr;
@@ -121,7 +123,14 @@ namespace Konclude {
 
 
 				CReuseCompletionGraphCache* CReuseCompletionGraphCache::writeExpandCache(CReuseCompletionGraphCacheEntryWriteData* writeData, CMemoryPool* memoryPools) {
+#ifdef WASM_BUILD
+					mWriteSyncMutex.lock();
+					CWriteCachedDataEvent* ev = new CWriteCachedDataEvent(writeData,memoryPools);
+					processCustomsEvents(CWriteCachedDataEvent::EVENTTYPE, ev);
+					mWriteSyncMutex.unlock();
+#else
 					postEvent(new CWriteCachedDataEvent(writeData,memoryPools));
+#endif
 					return this;
 				}
 
diff --git a/Source/Reasoner/Kernel/Cache/CReuseCompletionGraphCache.h b/Source/Reasoner/Kernel/Cache/CReuseCompletionGraphCache.h
index e46911ae..ec691570 100644
--- a/Source/Reasoner/Kernel/Cache/CReuseCompletionGraphCache.h
+++ b/Source/Reasoner/Kernel/Cache/CReuseCompletionGraphCache.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_KERNEL_CACHE_CREUSECOMPLETIONGRAPHCACHE_H
 
 // Libraries includes
-#include <QHash>
+#include "QtCompat.h"
 #include <QMutex>
 #include <QSemaphore>
 
@@ -108,6 +108,9 @@ namespace Konclude {
 						CReuseCompletionGraphCacheSlotItem* mSlotLinker;
 						CReuseCompletionGraphCacheReader* mReaderLinker;
 						QMutex mReaderSyncMutex;
+#ifdef WASM_BUILD
+						QMutex mWriteSyncMutex;
+#endif
 
 						CReuseCompletionGraphCompatibilityEntryHash* mEntyHash;
 						CCACHINGLIST<CReuseCompletionGraphCacheEntry*>* mEntyList;
diff --git a/Source/Reasoner/Kernel/Cache/CSaturationNodeAssociatedExpansionCache.cpp b/Source/Reasoner/Kernel/Cache/CSaturationNodeAssociatedExpansionCache.cpp
index 9c32883a..fb6ad0ea 100644
--- a/Source/Reasoner/Kernel/Cache/CSaturationNodeAssociatedExpansionCache.cpp
+++ b/Source/Reasoner/Kernel/Cache/CSaturationNodeAssociatedExpansionCache.cpp
@@ -44,7 +44,14 @@ namespace Konclude {
 						mConfAllowedNonDetExpansionCount = CConfigDataReader::readConfigInteger(config,"Konclude.Calculation.Optimization.SaturationExpansionSatisfiabilityCacheCount",1);
 					}
 
-					startThread(QThread::HighestPriority);
+					mConfDirectUpdateSynchronization = false;
+#ifdef WASM_BUILD
+					mConfDirectUpdateSynchronization = true;
+#endif
+
+					#ifndef WASM_BUILD
+						startThread(QThread::HighestPriority);
+					#endif
 				}
 
 
@@ -55,7 +62,16 @@ namespace Konclude {
 
 
 				CSaturationNodeAssociatedExpansionCache* CSaturationNodeAssociatedExpansionCache::writeCacheData(CSaturationNodeAssociatedExpansionCacheWriteData* writeData, CMemoryPool* memoryPools) {
-					postEvent(new CWriteSaturationCacheDataEvent(writeData,memoryPools));
+					if (mConfDirectUpdateSynchronization) {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::SatNodeAssocExpCache",logTr("[SatNodeAssocExpCache] sync writeCacheData"),this);
+						mDirectUpdateSyncMutex.lock();
+						CWriteSaturationCacheDataEvent* ev = new CWriteSaturationCacheDataEvent(writeData,memoryPools);
+						processCustomsEvents(EVENTWRITESATURATIONCACHEDATAENTRY, ev);
+						mDirectUpdateSyncMutex.unlock();
+					} else {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::SatNodeAssocExpCache",logTr("[SatNodeAssocExpCache] async writeCacheData"),this);
+						postEvent(new CWriteSaturationCacheDataEvent(writeData,memoryPools));
+					}
 					return this;
 				}
 
diff --git a/Source/Reasoner/Kernel/Cache/CSaturationNodeAssociatedExpansionCache.h b/Source/Reasoner/Kernel/Cache/CSaturationNodeAssociatedExpansionCache.h
index fc8463f5..7fd4348f 100644
--- a/Source/Reasoner/Kernel/Cache/CSaturationNodeAssociatedExpansionCache.h
+++ b/Source/Reasoner/Kernel/Cache/CSaturationNodeAssociatedExpansionCache.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_KERNEL_CACHE_CSATURATIONNODEASSOCIATEDEXPANSIONCACHE_H
 
 // Libraries includes
-#include <QHash>
+#include "QtCompat.h"
 #include <QMutex>
 #include <QSemaphore>
 
@@ -133,6 +133,9 @@ namespace Konclude {
 						CCacheStatistics mCacheStat;
 						CSaturationNodeAssociatedExpansionCacheContext mContext;
 
+						bool mConfDirectUpdateSynchronization;
+						QMutex mDirectUpdateSyncMutex;
+
 					// private methods
 					private:
 
diff --git a/Source/Reasoner/Kernel/Cache/CSignatureSatisfiableExpanderCache.cpp b/Source/Reasoner/Kernel/Cache/CSignatureSatisfiableExpanderCache.cpp
index e7321c2d..e8a8854c 100644
--- a/Source/Reasoner/Kernel/Cache/CSignatureSatisfiableExpanderCache.cpp
+++ b/Source/Reasoner/Kernel/Cache/CSignatureSatisfiableExpanderCache.cpp
@@ -58,7 +58,14 @@ namespace Konclude {
 					mNextCacheEntryRequiredSignatureRefCount = CConfigDataReader::readConfigInteger(config,"Konclude.Cache.SatisfiableExpander.RequiredSignatureReferenceCountIncrease",1);
 					mNextCacheEntryRequiredSignatureReferenceCountIncrease = CConfigDataReader::readConfigInteger(config,"Konclude.Cache.SatisfiableExpander.RequiredSignatureReferenceCountIncrease",1);
 
-					startThread(QThread::HighestPriority);
+					mConfDirectUpdateSynchronization = false;
+#ifdef WASM_BUILD
+					mConfDirectUpdateSynchronization = true;
+#endif
+
+					#ifndef WASM_BUILD
+						startThread(QThread::HighestPriority);
+					#endif
 				}
 
 
@@ -387,7 +394,16 @@ namespace Konclude {
 						cleanUnusedSlots(&mContext);
 					}
 #else 
-					postEvent(new CWriteCachedDataEvent(writeData,memoryPools));
+					if (mConfDirectUpdateSynchronization) {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::SigSatExpCache",logTr("[SigSatExpCache] sync writeCachedData"),this);
+						mDirectUpdateSyncMutex.lock();
+						CWriteCachedDataEvent* ev = new CWriteCachedDataEvent(writeData,memoryPools);
+						processCustomsEvents(EVENTWRITECACHEDDATAENTRY, ev);
+						mDirectUpdateSyncMutex.unlock();
+					} else {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::SigSatExpCache",logTr("[SigSatExpCache] async writeCachedData"),this);
+						postEvent(new CWriteCachedDataEvent(writeData,memoryPools));
+					}
 #endif
 					return this;
 				}
@@ -401,7 +417,16 @@ namespace Konclude {
 						cleanUnusedSlots(&mContext);
 					}
 #else 
-					postEvent(new CWriteExpandCachedEvent(prevSignature,newSignature,cacheValueList,depHash,memoryPools));
+					if (mConfDirectUpdateSynchronization) {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::SigSatExpCache",logTr("[SigSatExpCache] sync writeExpandCached"),this);
+						mDirectUpdateSyncMutex.lock();
+						CWriteExpandCachedEvent* ev = new CWriteExpandCachedEvent(prevSignature,newSignature,cacheValueList,depHash,memoryPools);
+						processCustomsEvents(EVENTWRITEEXPANDCACHEDENTRY, ev);
+						mDirectUpdateSyncMutex.unlock();
+					} else {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::SigSatExpCache",logTr("[SigSatExpCache] async writeExpandCached"),this);
+						postEvent(new CWriteExpandCachedEvent(prevSignature,newSignature,cacheValueList,depHash,memoryPools));
+					}
 #endif
 					return this;
 				}
@@ -410,7 +435,16 @@ namespace Konclude {
 #ifdef KONCLUCE_SATISFIBALE_EXPANDER_CACHE_DIRECT_WRITING
 					writeSatisfiableBranchedCachingData(&mContext,signature,cacheValueList,branchedList);
 #else 
-					postEvent(new CWriteSatisfiableBranchCachedEvent(signature,cacheValueList,branchedList,memoryPools));
+					if (mConfDirectUpdateSynchronization) {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::SigSatExpCache",logTr("[SigSatExpCache] sync writeSatisfiableBranchCached"),this);
+						mDirectUpdateSyncMutex.lock();
+						CWriteSatisfiableBranchCachedEvent* ev = new CWriteSatisfiableBranchCachedEvent(signature,cacheValueList,branchedList,memoryPools);
+						processCustomsEvents(EVENTWRITESATISFIABLEBRANCHCACHEDENTRY, ev);
+						mDirectUpdateSyncMutex.unlock();
+					} else {
+						LOG(INFO,"::Konclude::Reasoner::Kernel::Cache::SigSatExpCache",logTr("[SigSatExpCache] async writeSatisfiableBranchCached"),this);
+						postEvent(new CWriteSatisfiableBranchCachedEvent(signature,cacheValueList,branchedList,memoryPools));
+					}
 #endif
 					return this;
 				}
diff --git a/Source/Reasoner/Kernel/Cache/CSignatureSatisfiableExpanderCache.h b/Source/Reasoner/Kernel/Cache/CSignatureSatisfiableExpanderCache.h
index 0790cbb5..e2615d9b 100644
--- a/Source/Reasoner/Kernel/Cache/CSignatureSatisfiableExpanderCache.h
+++ b/Source/Reasoner/Kernel/Cache/CSignatureSatisfiableExpanderCache.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_KERNEL_CACHE_CSIGNATURESATISFIABLEEXPANDERCACHE_H
 
 // Libraries includes
-#include <QHash>
+#include "QtCompat.h"
 #include <QMutex>
 #include <QSemaphore>
 
@@ -162,6 +162,9 @@ namespace Konclude {
 						CSignatureSatisfiableExpanderCacheContext mContext;
 						CConfiguration* mConfig;
 
+						bool mConfDirectUpdateSynchronization;
+						QMutex mDirectUpdateSyncMutex;
+
 					// private methods
 					private:
 
